File: backend/src/Modules/Subscriptions/Domain/EntitlementSet.cs
namespace MoneyTracker.Modules.Subscriptions.Domain;

public sealed class EntitlementSet
{
    private static readonly IReadOnlySet<FeatureKey> PremiumFeatures = new HashSet<FeatureKey>
    {
        FeatureKey.BankSync,
        FeatureKey.PremiumInsights,
        FeatureKey.UnlimitedBudgets,
        FeatureKey.UnlimitedBillReminders,
        FeatureKey.ExportData
    };

    private static readonly IReadOnlySet<FeatureKey> FreeFeatures =
        new HashSet<FeatureKey>();

    private EntitlementSet(SubscriptionTier tier, IReadOnlySet<FeatureKey> featureKeys)
    {
        Tier = tier;
        FeatureKeys = featureKeys;
    }

    public SubscriptionTier Tier { get; }

    public IReadOnlySet<FeatureKey> FeatureKeys { get; }

    public bool HasFeature(FeatureKey feature) => FeatureKeys.Contains(feature);

    public static EntitlementSet ForTier(SubscriptionTier tier) => tier switch
    {
        SubscriptionTier.Premium => new EntitlementSet(tier, PremiumFeatures),
        SubscriptionTier.Trial => new EntitlementSet(tier, PremiumFeatures),
        SubscriptionTier.Free => new EntitlementSet(tier, FreeFeatures),
        _ => new EntitlementSet(SubscriptionTier.Free, FreeFeatures)
    };
}

File: mobile/lib/features/subscriptions/domain/entitlement_set.dart
import 'feature_key.dart';
import 'subscription_tier.dart';

class EntitlementSet {
  const EntitlementSet._({
    required this.tier,
    required this.featureKeys,
    this.trialExpiresAtUtc,
    this.currentPeriodEndUtc,
  });

  final SubscriptionTier tier;
  final Set<FeatureKey> featureKeys;
  final DateTime? trialExpiresAtUtc;
  final DateTime? currentPeriodEndUtc;

  bool hasFeature(FeatureKey feature) => featureKeys.contains(feature);

  factory EntitlementSet.forTier(SubscriptionTier tier) {
    switch (tier) {
      case SubscriptionTier.premium:
      case SubscriptionTier.trial:
        return EntitlementSet._(
          tier: tier,
          featureKeys: const {
            FeatureKey.bankSync,
            FeatureKey.premiumInsights,
            FeatureKey.unlimitedBudgets,
            FeatureKey.unlimitedBillReminders,
            FeatureKey.exportData,
          },
        );
      case SubscriptionTier.free:
        return EntitlementSet._(
          tier: tier,
          featureKeys: const {},
        );
    }
  }

  factory EntitlementSet.fromApiResponse({
    required String tier,
    required List<String> featureKeys,
    DateTime? trialExpiresAtUtc,
    DateTime? currentPeriodEndUtc,
  }) {
    final parsedTier = SubscriptionTier.fromString(tier);
    final parsedFeatures = <FeatureKey>{};
    for (final key in featureKeys) {
      final parsed = FeatureKey.fromString(key);
      if (parsed != null) {
        parsedFeatures.add(parsed);
      }
    }

    return EntitlementSet._(
      tier: parsedTier,
      featureKeys: parsedFeatures,
      trialExpiresAtUtc: trialExpiresAtUtc,
      currentPeriodEndUtc: currentPeriodEndUtc,
    );
  }

  factory EntitlementSet.free() {
    return EntitlementSet.forTier(SubscriptionTier.free);
  }
}
